In [164]:
# Install a pip package in the current Jupyter kernel
import sys
!{sys.executable} -m pip install pygithub
!{sys.executable} -m pip install gitpython
!conda install --yes --prefix {sys.prefix} git


Solving environment: ...working... done

# All requested packages already installed.



In [165]:
import numpy as np
import pandas as pd
import os

# Specify git executable file for GitPython in Jupyter Notebook (In IDE, it can still work without this line.)
os.environ["GIT_PYTHON_GIT_EXECUTABLE"] = "C:\Program Files\Git\cmd\git.exe"

import git
from git import RemoteProgress

from git import Repo
import matplotlib.pyplot as plt
import seaborn as sns
%matplotlib inline

class Progress(RemoteProgress):
    def update(self, op_code, cur_count, max_count=None, message=''):
        print(self._cur_line)
        
remote_link = "https://github.com/apache/groovy"
local_link = "repo/apache"
# Uncomment to clone
Repo.clone_from(remote_link, local_link, progress=Progress())

repo = Repo(local_link)
fixing_commit = "4df8b652aa018a5d5d1cda8fba938bf3422db31c"
affected_file = "src/main/org/codehaus/groovy/runtime/MethodClosure.java"

remote: Counting objects:   6% (1/15)
remote: Counting objects:  13% (2/15)
remote: Counting objects:  20% (3/15)
remote: Counting objects:  26% (4/15)
remote: Counting objects:  33% (5/15)
remote: Counting objects:  40% (6/15)
remote: Counting objects:  46% (7/15)
remote: Counting objects:  53% (8/15)
remote: Counting objects:  60% (9/15)
remote: Counting objects:  66% (10/15)
remote: Counting objects:  73% (11/15)
remote: Counting objects:  80% (12/15)
remote: Counting objects:  86% (13/15)
remote: Counting objects:  93% (14/15)
remote: Counting objects: 100% (15/15)
remote: Counting objects: 100% (15/15), done.
remote: Compressing objects:   7% (1/14)
remote: Compressing objects:  14% (2/14)
remote: Compressing objects:  21% (3/14)
remote: Compressing objects:  28% (4/14)
remote: Compressing objects:  35% (5/14)
remote: Compressing objects:  42% (6/14)
remote: Compressing objects:  50% (7/14)
remote: Compressing objects:  57% (8/14)
remote: Compressing objects:  64% (9/14)
remote: C

Receiving objects:  25% (79138/313017), 95.01 MiB | 1.82 MiB/s
Receiving objects:  25% (79610/313017), 96.78 MiB | 1.72 MiB/s
Receiving objects:  25% (80632/313017), 98.39 MiB | 1.62 MiB/s
Receiving objects:  25% (80632/313017), 99.65 MiB | 1.56 MiB/s
Receiving objects:  25% (80632/313017), 100.86 MiB | 1.46 MiB/s
Receiving objects:  26% (81385/313017), 100.86 MiB | 1.46 MiB/s
Receiving objects:  26% (83811/313017), 101.45 MiB | 1.41 MiB/s
Receiving objects:  27% (84515/313017), 101.76 MiB | 1.28 MiB/s
Receiving objects:  27% (85579/313017), 102.58 MiB | 1.10 MiB/s
Receiving objects:  27% (85579/313017), 103.64 MiB | 1002.00 KiB/s
Receiving objects:  27% (85579/313017), 104.50 MiB | 955.00 KiB/s
Receiving objects:  27% (85579/313017), 105.25 MiB | 846.00 KiB/s
Receiving objects:  27% (85579/313017), 106.22 MiB | 918.00 KiB/s
Receiving objects:  27% (85579/313017), 107.58 MiB | 999.00 KiB/s
Receiving objects:  27% (85579/313017), 108.58 MiB | 997.00 KiB/s
Receiving objects:  27% (85580/

Receiving objects:  97% (304788/313017), 173.66 MiB | 1.61 MiB/s
Receiving objects:  98% (306757/313017), 174.52 MiB | 1.61 MiB/s
Receiving objects:  99% (309887/313017), 174.52 MiB | 1.61 MiB/s
Receiving objects:  99% (312429/313017), 174.52 MiB | 1.61 MiB/s
Receiving objects: 100% (313017/313017), 175.30 MiB | 1.57 MiB/s
Receiving objects: 100% (313017/313017), 175.44 MiB | 1.24 MiB/s, done.
Resolving deltas:   0% (0/195108)
Resolving deltas:   1% (2120/195108)
Resolving deltas:   2% (3945/195108)
Resolving deltas:   3% (5868/195108)
Resolving deltas:   4% (7819/195108)
Resolving deltas:   5% (9765/195108)
Resolving deltas:   6% (11793/195108)
Resolving deltas:   7% (13695/195108)
Resolving deltas:   8% (15701/195108)
Resolving deltas:   9% (17566/195108)
Resolving deltas:  10% (19562/195108)
Resolving deltas:  11% (21474/195108)
Resolving deltas:  12% (23458/195108)
Resolving deltas:  13% (25371/195108)
Resolving deltas:  14% (27323/195108)
Resolving deltas:  15% (29305/195108)
Reso

In [166]:
#a.	What was the message and title of the fixing commit? 
#Was there any mention of fixing a bug or vulnerability? 
#Hint: git diff or git show
show_data = repo.git.show("-s",fixing_commit).splitlines()
for line in show_data:
    print(line)

commit 4df8b652aa018a5d5d1cda8fba938bf3422db31c
Author: Jochen Theodorou <blackdrag@gmx.org>
Date:   Mon Oct 3 18:22:28 2016 +0200

    fix possible deserialization exploit by overriding readObject


In [167]:
#b.	How many total files were affected in the fixing commit? 
#Hint: git diff or git show
diff_data = repo.git.diff(fixing_commit + "^", fixing_commit).splitlines()
for line in diff_data:
    print(line)

diff --git a/src/main/org/codehaus/groovy/runtime/MethodClosure.java b/src/main/org/codehaus/groovy/runtime/MethodClosure.java
index 6dfb2f3010..cacdca8421 100644
--- a/src/main/org/codehaus/groovy/runtime/MethodClosure.java
+++ b/src/main/org/codehaus/groovy/runtime/MethodClosure.java
@@ -21,6 +21,7 @@ package org.codehaus.groovy.runtime;
 import groovy.lang.Closure;
 import groovy.lang.MetaMethod;
 
+import java.io.IOException;
 import java.util.List;
 
 
@@ -71,6 +72,13 @@ public class MethodClosure extends Closure {
         }
         throw new UnsupportedOperationException();
     }
+
+    private void readObject(java.io.ObjectInputStream stream) throws IOException, ClassNotFoundException {
+        if (ALLOW_RESOLVE) {
+            stream.defaultReadObject();
+        }
+        throw new UnsupportedOperationException();
+    }
     
     public Object getProperty(String property) {
         if ("method".equals(property)) {


In [168]:
#c.	How many total directories were affected in the fixing commit? 
#For example, if a file path is: abc/def/File.java, 
#then its directory is abc/def. Hint: git diff or git show
diff_data = repo.git.diff(fixing_commit + "^", fixing_commit).splitlines()
for line in diff_data:
    print(line)

diff --git a/src/main/org/codehaus/groovy/runtime/MethodClosure.java b/src/main/org/codehaus/groovy/runtime/MethodClosure.java
index 6dfb2f3010..cacdca8421 100644
--- a/src/main/org/codehaus/groovy/runtime/MethodClosure.java
+++ b/src/main/org/codehaus/groovy/runtime/MethodClosure.java
@@ -21,6 +21,7 @@ package org.codehaus.groovy.runtime;
 import groovy.lang.Closure;
 import groovy.lang.MetaMethod;
 
+import java.io.IOException;
 import java.util.List;
 
 
@@ -71,6 +72,13 @@ public class MethodClosure extends Closure {
         }
         throw new UnsupportedOperationException();
     }
+
+    private void readObject(java.io.ObjectInputStream stream) throws IOException, ClassNotFoundException {
+        if (ALLOW_RESOLVE) {
+            stream.defaultReadObject();
+        }
+        throw new UnsupportedOperationException();
+    }
     
     public Object getProperty(String property) {
         if ("method".equals(property)) {


In [169]:
#d.	How many total lines of code (including comments and blank lines) \
#were deleted? Hint: git diff or git show
#e.	How many total lines of code (including comments and blank lines) 
#were added? Hint: git diff or git show
del_data = repo.git.log("--numstat","--oneline",fixing_commit).splitlines()
for line in del_data:
    print(line)

4df8b652aa fix possible deserialization exploit by overriding readObject
8	0	src/main/org/codehaus/groovy/runtime/MethodClosure.java
675a04eab7 GROOVY-7930: @SelfType in hierarchy with an interface in the middle (closes #428)
1	1	src/main/org/codehaus/groovy/ast/tools/GeneralUtils.java
16	0	src/test/org/codehaus/groovy/transform/traitx/TraitASTTransformationTest.groovy
da30fa2005 GROOVY-7946 - allow writable values for StreamingJsonBuilder (closes #429)
21	3	subprojects/groovy-json/src/main/java/groovy/json/StreamingJsonBuilder.java
24	0	subprojects/groovy-json/src/test/groovy/groovy/json/StreamingJsonBuilderTest.groovy
e36835d58a GROOVY-7933: fix method selection for boolean and char primitives (closes #424)
42	22	src/main/org/codehaus/groovy/runtime/MetaClassHelper.java
48	0	src/test/groovy/bugs/Groovy7933Bug.groovy
1	1	src/test/org/codehaus/groovy/transform/CanonicalComponentsTransformTest.groovy
a982938330 GROOVY-7935: fix style guide example for required parentheses
3	4	src/spec/d

1	1	src/test/org/codehaus/groovy/runtime/JdkDynamicProxyTest.java
1	1	src/test/org/codehaus/groovy/runtime/MetaClassHelperTest.java
1	1	src/test/org/codehaus/groovy/runtime/MethodFailureTest.java
1	1	src/test/org/codehaus/groovy/runtime/MethodKeyTest.java
1	1	src/test/org/codehaus/groovy/runtime/MethodRankHelperTest.java
1	1	src/test/org/codehaus/groovy/runtime/MockGroovyObject.java
1	1	src/test/org/codehaus/groovy/runtime/NewStaticMetaMethodTest.java
1	1	src/test/org/codehaus/groovy/runtime/PropertyTest.java
1	1	src/test/org/codehaus/groovy/runtime/StaticInitTest.java
1	1	src/test/org/codehaus/groovy/runtime/StringGroovyMethodsTest.java
1	1	src/test/org/codehaus/groovy/runtime/TupleListTest.java
1	1	src/test/org/codehaus/groovy/runtime/m12n/Groovy6496Extension.java
1	1	src/test/org/codehaus/groovy/runtime/m12n/TestStaticStringExtension.java
1	1	src/test/org/codehaus/groovy/runtime/m12n/TestStringExtension.java
1	1	src/test/org/codehaus/groovy/syntax/TokenTest.java
1	1	src/test/org/cod

3	16	subprojects/groovy-servlet/src/spec/doc/servlet-userguide.adoc
3	16	subprojects/groovy-sql/src/spec/doc/sql-userguide.adoc
3	16	subprojects/groovy-swing/src/spec/doc/swing-builder.adoc
3	16	subprojects/groovy-templates/src/spec/doc/markup-template-engine.adoc
3	16	subprojects/groovy-templates/src/spec/doc/template-engines.adoc
3	16	subprojects/groovy-xml/src/spec/doc/dom-builder.adoc
3	16	subprojects/groovy-xml/src/spec/doc/sax-builder.adoc
3	16	subprojects/groovy-xml/src/spec/doc/stax-builder.adoc
3	16	subprojects/groovy-xml/src/spec/doc/xml-userguide.adoc
e7bdfd9672 Apache license headers for gradle files
1	1	build.gradle
0	1	config/licensing/rat-excludes.txt
1	1	settings.gradle
1	1	subprojects/groovy-ant/build.gradle
1	1	subprojects/groovy-bsf/build.gradle
1	1	subprojects/groovy-console/build.gradle
1	1	subprojects/groovy-docgenerator/build.gradle
1	1	subprojects/groovy-groovydoc/build.gradle
1	1	subprojects/groovy-groovysh/build.gradle
1	1	subprojects/groovy-jmx/build.gradle
1

14	11	src/test/groovy/txn/TransactionBean.java
14	11	src/test/groovy/txn/TransactionBuilder.java
14	11	src/test/groovy/txn/TransactionTest.groovy
14	11	src/test/groovy/ui/GroovyMainTest.groovy
14	11	src/test/groovy/util/BuilderSupportTest.groovy
14	10	src/test/groovy/util/CliBuilderTest.groovy
14	11	src/test/groovy/util/ConfigObjectTest.groovy
14	11	src/test/groovy/util/ConfigSlurperTest.groovy
14	11	src/test/groovy/util/DelegatingScriptTest.groovy
14	11	src/test/groovy/util/EvalTest.java
14	11	src/test/groovy/util/FactoryBuilderSupportTest.groovy
14	12	src/test/groovy/util/FileTreeBuilderTest.groovy
14	11	src/test/groovy/util/GroovyCollectionsStarImportTest.groovy
14	11	src/test/groovy/util/GroovyCollectionsTest.groovy
14	11	src/test/groovy/util/GroovyScriptEngineReloadingTest.groovy
14	11	src/test/groovy/util/HeadlessTestSupport.groovy
14	11	src/test/groovy/util/IndentPrinterTest.groovy
14	11	src/test/groovy/util/JavadocAssertionTestBuilderTest.groovy
14	11	src/test/groovy/util/MiscS

35	0	src/spec/test-resources/scoping.groovy
40	0	src/spec/test-resources/scoping_alt.groovy
32	0	src/spec/test-resources/selfcheck.groovy
21	0	src/spec/test-resources/setup.groovy
24	0	src/spec/test-resources/unresolvedattribute.groovy
27	0	src/spec/test-resources/unresolvedproperty.groovy
24	0	src/spec/test-resources/unresolvedvariable.groovy
40	0	src/spec/test/asciidoctor/Utils.groovy
33	0	src/spec/test/typing/PrecompiledExtension.groovy
44	0	src/spec/test/typing/PrecompiledJavaExtension.java
23	0	src/spec/test/typing/Robot.groovy
467	0	src/spec/test/typing/TypeCheckingExtensionSpecTest.groovy
a6ebe19147 GROOVY-7309: Incorrect args processing in GroovyShell.run(Reader, String, List)
1	1	src/main/groovy/lang/GroovyShell.java
432ad401e5 GROOVY-7309: Incorrect args processing in GroovyShell.run(Reader, String, List)
2	2	src/main/groovy/lang/GroovyShell.java
da52fee28c correct test name
1	1	src/test/org/codehaus/groovy/transform/DelegateTransformTest.groovy
2b027937c8 GROOVY-7299: add a 

28a2230944 minor refactor: typo in template
1	1	subprojects/groovy-docgenerator/src/main/resources/org/codehaus/groovy/tools/template.class.html
e785864794 GROOVY-6753: groovysh: better error output for loading file with hyphens
1	1	subprojects/groovy-groovysh/src/main/groovy/org/codehaus/groovy/tools/shell/commands/DisplayCommand.groovy
1	1	subprojects/groovy-groovysh/src/main/groovy/org/codehaus/groovy/tools/shell/commands/EditCommand.groovy
2	2	subprojects/groovy-groovysh/src/main/groovy/org/codehaus/groovy/tools/shell/commands/LoadCommand.groovy
4	4	subprojects/groovy-groovysh/src/main/groovy/org/codehaus/groovy/tools/shell/commands/RecordCommand.groovy
2	2	subprojects/groovy-groovysh/src/main/groovy/org/codehaus/groovy/tools/shell/commands/SaveCommand.groovy
8ea89d05d8 Merge pull request #404 from nobeans/wk/tweak-generate-gdk-doc
414a02820c GROOVY-6749: tweaked a layout and style of template.class.html
17	3	subprojects/groovy-docgenerator/src/main/resources/org/codehaus/groovy/to

90e8508694 Merge pull request #308 from tkruse/groovy6395
d9f4d2b427 Merge branch 'master' into GROOVY-5924
ebe224cd86 GROOVY-6395: Complete navigable properties for Map and Node instances
64	0	subprojects/groovy-groovysh/src/main/groovy/org/codehaus/groovy/tools/shell/completion/NavigablePropertiesCompleter.groovy
14	3	subprojects/groovy-groovysh/src/main/groovy/org/codehaus/groovy/tools/shell/completion/ReflectionCompletor.groovy
39	0	subprojects/groovy-groovysh/src/test/groovy/org/codehaus/groovy/tools/shell/completion/NavigablePropertiesCompleterTest.groovy
3	1	subprojects/groovy-groovysh/src/test/groovy/org/codehaus/groovy/tools/shell/completion/ReflectionCompletorTest.groovy
603238318f exclude nio subproject when using java 5 or 6
8	3	settings.gradle
9fc83c7f30 fixed tests (at least under Windows) and unified formating
50	26	subprojects/groovy-nio/src/test/groovy/NioGroovyMethodsTest.groovy
1ff0159230 Try to fix nio module build
11	11	subprojects/groovy-nio/src/test/groovy/NioGro

1	12	src/main/org/codehaus/groovy/transform/BaseScriptASTTransformation.java
4873deb30d annotation to set base script of the script
58	0	src/main/groovy/transform/BaseScript.java
116	0	src/main/org/codehaus/groovy/transform/BaseScriptASTTransformation.java
131	0	src/test/org/codehaus/groovy/transform/BaseScriptTransformTest.groovy
7b30382182 GROOVY-6180: TypeCheckingMode.SKIP not always applied to anonymous inner classes
1	1	src/main/org/codehaus/groovy/transform/sc/StaticCompilationVisitor.java
1	1	src/main/org/codehaus/groovy/transform/stc/StaticTypeCheckingVisitor.java
23	0	src/test/groovy/transform/stc/TypeCheckingModeTest.groovy
24	0	src/test/org/codehaus/groovy/classgen/asm/sc/TypeCheckingModeStaticCompileTest.groovy
89934b3c7f Improve error message when operand stack is empty to give more context and give the user/debugger a better chance to find out what code is causing trouble.
19	4	src/main/org/codehaus/groovy/classgen/asm/OperandStack.java
f32b3d9cb2 GROOVY-6152: LazyList do

c96c22290a Merge remote-tracking branch 'remotes/upstream/master' into stc-refactor
5960aabbbc GROOVY-5767: Improve DSL type checking    * make use of a class node stack too    * minor refactoring of assignment tracking
32	40	src/main/org/codehaus/groovy/transform/stc/StaticTypeCheckingVisitor.java
53	2	src/main/org/codehaus/groovy/transform/stc/TypeCheckingContext.java
49c8c049ec GROOVY-5767: Improve DSL type checking    * more refactoring: use only one return adder which is used for both closures and methods    * make type checking context aware of an enclosing closure stack    * make type checking context aware of an enclosing method stack    * make type checking context aware of a binary expression stack    * temporarily replace plugin with an error handler
75	0	src/main/org/codehaus/groovy/transform/stc/DefaultTypeCheckingErrorHandler.java
2	1	src/main/org/codehaus/groovy/transform/stc/SignatureCodec.java
2	1	src/main/org/codehaus/groovy/transform/stc/SignatureCodecVersion1.java
1

3	3	src/test/groovy/lang/MixinTest.groovy
26e9a5b630 replaced tabs with spaces
9	9	benchmark/bench/ary.groovy
2	2	config/codenarc/codenarc.groovy
21	21	src/main/groovy/swing/factory/CellEditorFactory.groovy
31	31	src/main/groovy/swing/factory/ColumnFactory.groovy
96	96	src/main/groovy/swing/impl/ClosureCellEditor.groovy
1	1	src/main/groovy/ui/Console.groovy
3	3	src/main/org/codehaus/groovy/tools/shell/ReflectionCompletor.groovy
27	27	src/test/groovy/bugs/Groovy5030Bug.groovy
3	3	src/test/groovy/bugs/Groovy5033Bug.groovy
38	38	src/test/groovy/bugs/Groovy5056Bug.groovy
18	18	src/test/groovy/bugs/Groovy5193Bug.groovy
27	27	src/test/groovy/bugs/VerifyErrorBug.groovy
1	1	src/test/org/codehaus/groovy/control/customizers/SecureASTCustomizerTest.groovy
d05459b740 replaced tabs with spaces
4	4	benchmark/bench/ackermann.java
9	9	benchmark/bench/ary.java
63	63	benchmark/bench/binarytrees.java
123	123	benchmark/bench/chameneos.java
49	49	benchmark/bench/echo.java
29	29	benchmark/bench/except.java


2	1	src/main/org/codehaus/groovy/classgen/Verifier.java
13	1	src/main/org/codehaus/groovy/transform/StaticTypesTransformation.java
137	0	src/test/groovy/transform/stc/ReturnsSTCTest.groovy
06eef42e88 Methods from the String class should be considered valid on GString too in static type checking
1	0	src/main/org/codehaus/groovy/transform/StaticTypesTransformation.java
8	0	src/test/groovy/transform/stc/TypeInferenceSTCTest.groovy
4af635e51b Fixed case where no constructor is found in static type checking
12	2	src/main/org/codehaus/groovy/transform/StaticTypesTransformation.java
0	1	src/test/groovy/transform/stc/DefaultGroovyMethodsSTCTest.groovy
43c1e9a41d Added some unit tests for assignment in static type checking
12	3	src/main/org/codehaus/groovy/transform/StaticTypesTransformation.java
35	0	src/test/groovy/transform/stc/STCAssignmentTest.groovy
6de942d2f3 Implemented basic flow analysis for assignments in static type checking
51	33	src/main/org/codehaus/groovy/transform/StaticTypesTr

18823a326f added @since javadoc tags and upgraded some javadocs on the AST transformations.
1	0	src/main/groovy/transform/Field.java
1	0	src/main/groovy/transform/Immutable.java
1	0	src/main/groovy/transform/IndexedProperty.java
1	0	src/main/groovy/transform/InheritConstructors.java
1	0	src/main/groovy/transform/Synchronized.java
4	1	src/main/groovy/transform/WithReadLock.java
4	1	src/main/groovy/transform/WithWriteLock.java
15c14202a6 Added Javadoc to TracingInterceptor and BenchmarkInterceptor, which previously had none.
29	2	src/main/groovy/lang/BenchmarkInterceptor.java
33	0	src/main/groovy/lang/TracingInterceptor.java
b97ad9918c GROOVY-4516 Fixed handling of params with defaults in @Delegate
1	0	src/main/org/codehaus/groovy/transform/DelegateASTTransformation.java
36	0	src/test/groovy/bugs/Groovy4516Bug.groovy
4d975c1610 GROOVY-4525: Sql.newInstance map version assumes map values are strings
12	8	src/main/groovy/sql/Sql.java
2	1	src/test/groovy/sql/TestHelper.groovy
8c86e09a8f GRO

15	0	src/main/org/codehaus/groovy/runtime/dgmimpl/arrays/CharacterArrayGetAtMetaMethod.java
15	0	src/main/org/codehaus/groovy/runtime/dgmimpl/arrays/DoubleArrayGetAtMetaMethod.java
5	3	src/main/org/codehaus/groovy/runtime/metaclass/ConcurrentReaderHashMap.java
24	30	src/main/org/codehaus/groovy/runtime/metaclass/MemoryAwareConcurrentReadMap.java
15	0	src/main/org/codehaus/groovy/runtime/metaclass/MixinInstanceMetaMethod.java
15	0	src/main/org/codehaus/groovy/runtime/metaclass/MixinInstanceMetaProperty.java
15	0	src/main/org/codehaus/groovy/runtime/typehandling/NumberMathModificationInfo.java
0	1	src/main/org/codehaus/groovy/syntax/RuntimeParserException.java
15	0	src/main/org/codehaus/groovy/tools/StringHelper.java
15	0	src/main/org/codehaus/groovy/tools/groovydoc/LinkArgument.java
15	0	src/main/org/codehaus/groovy/tools/groovydoc/SimpleGroovyTag.java
15	0	src/main/org/codehaus/groovy/tools/groovydoc/gstringTemplates/GroovyDocTemplateInfo.java
15	0	src/main/org/codehaus/groovy/tools/gs

58	24	src/main/groovy/sql/Sql.java
83b6c0aa3e GROOVY-3477: Add setVariables method to GroovyServlet for easy extending of binded vars in custom classes (merge rev 16026)
60	3	src/main/groovy/servlet/GroovyServlet.java
1d36071b60 comment typos
7	7	src/main/groovy/grape/Grape.java
2	2	src/main/groovy/grape/GrapeIvy.groovy
9172f1be01 GROOVY-3474 Fixed the recursion error in findMixinMethod() that was causing the StackOverflowError.
2	1	src/main/groovy/lang/ExpandoMetaClass.java
22	0	src/test/groovy/lang/MixinTest.groovy
7c43bf3272 GROOVY-3470 - fix bind validator
2	2	src/main/org/codehaus/groovy/binding/AbstractFullBinding.java
58	3	src/test/groovy/swing/SwingBuilderBindingsTest.groovy
ba22845671 GROOVY-3465 Statically imported methods, if called with named parameters, were not getting resolved correctly. Fixed it.
3	0	src/main/org/codehaus/groovy/ast/ClassNode.java
14	0	src/test/groovy/bugs/Groovy3465Bug.groovy
10	0	src/test/groovy/bugs/Groovy3465Helper.groovy
318e3b2a05 GROOVY-3461: rem

1	1	src/main/groovy/util/XmlNodePrinter.java
10	0	src/test/groovy/util/XmlNodePrinterTest.groovy
700d847323 JIRA 3028
12	9	src/main/org/codehaus/groovy/runtime/DefaultGroovyMethods.java
18	15	src/test/groovy/RegularExpressionsTest.groovy
dde8e55f85 JIRA 3028, Make Regular Expression handling consistent
43	67	src/main/org/codehaus/groovy/runtime/DefaultGroovyMethods.java
127	15	src/test/groovy/RegularExpressionsTest.groovy
94017baaea merging rev 13427 and rev 13428 to trunk (was a fix resulted from GROOVY-3021 and the broken grails build for 1.5.x)
114	87	src/main/org/codehaus/groovy/classgen/Verifier.java
62	0	src/test/gls/invocation/GroovyObjectInheritanceTest.groovy
8408afb8ec GROOVY-3021 remove old workarounds
0	19	src/main/groovy/swing/SwingBuilder.groovy
df98cb4a55 MetaBuilderTest
266	0	src/test/org/codehaus/groovy/tools/MetaBuilderTest.groovy
05258ffc72 test case for GROOVY-3021
38	0	src/test/gls/invocation/GroovyObjectInheritanceTest.groovy
13978e0aa6 fix spread dot operator for

191	0	Groovy-Dev/src/main/groovy/servlet/GroovyServlet.java
199	0	Groovy-Dev/src/main/groovy/servlet/ServletBinding.java
92	0	Groovy-Dev/src/main/groovy/servlet/ServletCategory.java
488	0	Groovy-Dev/src/main/groovy/servlet/TemplateServlet.java
10	0	Groovy-Dev/src/main/groovy/servlet/package.html
53	0	Groovy-Dev/src/main/groovy/sql/CallResultSet.java
277	0	Groovy-Dev/src/main/groovy/sql/DataSet.java
27	0	Groovy-Dev/src/main/groovy/sql/ExpandedVariable.java
90	0	Groovy-Dev/src/main/groovy/sql/GroovyResultSet.java
263	0	Groovy-Dev/src/main/groovy/sql/GroovyResultSetExtension.java
120	0	Groovy-Dev/src/main/groovy/sql/GroovyResultSetProxy.java
164	0	Groovy-Dev/src/main/groovy/sql/GroovyRowResult.java
26	0	Groovy-Dev/src/main/groovy/sql/InOutParameter.java
37	0	Groovy-Dev/src/main/groovy/sql/InParameter.java
30	0	Groovy-Dev/src/main/groovy/sql/OutParameter.java
26	0	Groovy-Dev/src/main/groovy/sql/ResultSetOutParameter.java
1490	0	Groovy-Dev/src/main/groovy/sql/Sql.java
43	0	Groovy-Dev/src/ma

IOPub data rate exceeded.
The notebook server will temporarily stop sending output
to the client in order to avoid crashing it.
To change this limit, set the config variable
`--NotebookApp.iopub_data_rate_limit`.

Current values:
NotebookApp.iopub_data_rate_limit=1000000.0 (bytes/sec)
NotebookApp.rate_limit_window=3.0 (secs)



0	15	src/test/groovy/bugs/StaticMethodSamePrivate.groovy
0	9	src/test/groovy/bugs/StaticMethodSameProtected.groovy
0	9	src/test/groovy/bugs/priv/StaticMethodPackaged.groovy
0	15	src/test/groovy/bugs/priv/StaticMethodPrivate.groovy
0	9	src/test/groovy/bugs/priv/StaticMethodProtected.groovy
0	9	src/test/groovy/bugs/priv/StaticMethodPublic.groovy
e4ab1d6f05 Trying to roll back the changes to invokeMethod - hold on to your hats!
0	12	src/main/groovy/lang/Closure.java
0	6	src/main/groovy/lang/GroovyObject.java
0	4	src/main/groovy/lang/GroovyObjectSupport.java
1	1	src/main/groovy/lang/GroovyShell.java
18	276	src/main/groovy/lang/MetaClass.java
0	85	src/main/groovy/lang/MetaMethod.java
0	17	src/main/groovy/lang/ProxyMetaClass.java
0	24	src/main/groovy/lang/Script.java
1	17	src/main/groovy/lang/Sequence.java
0	14	src/main/groovy/mock/GroovyMock.java
1	1	src/main/groovy/text/GStringTemplateEngine.java
0	96	src/main/groovy/util/BuilderSupport.java
0	19	src/main/groovy/util/Expando.java
0	5	src/m

2	1	src/test/groovy/util/BuilderSupportTest.groovy
6bc124b9db more tests eat dirt
3	0	src/main/groovy/lang/GroovyShell.java
0	1	src/test/UberTestCase2.java
ce75db1c82 more tests fixed
56	57	src/main/org/codehaus/groovy/sandbox/markup/AbstractStreamingBuilder.groovy
138	139	src/main/org/codehaus/groovy/sandbox/markup/StreamingDOMBuilder.groovy
108	109	src/main/org/codehaus/groovy/sandbox/markup/StreamingMarkupBuilder.groovy
116	117	src/main/org/codehaus/groovy/sandbox/markup/StreamingSAXBuilder.groovy
1	1	src/test/UberTestCase2.java
1	1	src/test/groovy/bugs/IanMaceysBug.java
a3553feaab fixed a bunch more test cases
3	1	src/main/org/codehaus/groovy/antlr/AntlrParserPlugin.java
6	6	src/test/UberTestCase.java
1	1	src/test/groovy/SafeNavigationTest.groovy
2	2	src/test/groovy/bugs/MarkupAndMethodBug.groovy
1	1	src/test/groovy/bugs/SuperMethod2Bug.groovy
4	3	src/test/groovy/bugs/VariablePrecedence.groovy
09f68032d5 more bugs bite the dust
2	2	src/test/UberTestCase.java
2	1	src/test/groovy/Lis

17	0	src/test-new/groovy/bugs/ConstructorParameterBug.groovy
19	0	src/test-new/groovy/bugs/DefVariableBug.groovy
70	0	src/test-new/groovy/bugs/DoubleSizeParametersBug.groovy
38	0	src/test-new/groovy/bugs/ForAndSqlBug.groovy
62	0	src/test-new/groovy/bugs/ForLoopBug.groovy
10	0	src/test-new/groovy/bugs/FullyQualifiedClassBug.groovy
65	0	src/test-new/groovy/bugs/GetterBug.groovy
40	0	src/test-new/groovy/bugs/Groovy239_Bug.groovy
63	0	src/test-new/groovy/bugs/Groovy249_Bug.groovy
36	0	src/test-new/groovy/bugs/Groovy252_Bug.groovy
21	0	src/test-new/groovy/bugs/Groovy278_Bug.groovy
47	0	src/test-new/groovy/bugs/Groovy303_Bug.groovy
26	0	src/test-new/groovy/bugs/Groovy308_Bug.groovy
21	0	src/test-new/groovy/bugs/Groovy389_Bug.groovy
16	0	src/test-new/groovy/bugs/Groovy513_Bug.groovy
45	0	src/test-new/groovy/bugs/Groovy558_616_Bug.groovy
11	0	src/test-new/groovy/bugs/Groovy666_Bug.groovy
11	0	src/test-new/groovy/bugs/GuillaumesBug.groovy
40	0	src/test-new/groovy/bugs/GuillaumesMapBug.groovy
67

8	2	src/main/groovy/lang/GroovyShell.java
891ad59d07 removed some old cruft we no longer use.
1	0	.cvsignore
12	0	.project
18	5	src/main/org/codehaus/groovy/ast/ClassNode.java
86	0	src/main/org/codehaus/groovy/ast/ImportNode.java
10	0	src/main/org/codehaus/groovy/ast/MetadataNode.java
9	3	src/main/org/codehaus/groovy/ast/ModuleNode.java
1	1	src/main/org/codehaus/groovy/ast/PropertyNode.java
10	5	src/main/org/codehaus/groovy/classgen/Verifier.java
0	15	src/main/org/codehaus/groovy/syntax/SemanticException.java
2	0	src/main/org/codehaus/groovy/syntax/SyntaxException.java
0	82	src/main/org/codehaus/groovy/syntax/parser/SemanticVerificationResult.java
0	14	src/main/org/codehaus/groovy/syntax/parser/SemanticVerifier.java
0	11	src/main/org/codehaus/groovy/tools/Compiler.java
403ef4b99c updated the documention to include Stephan's patch
30	12	xdocs/embedding.wiki
e02b39975d Refactored GroovyClassLoader.parseClass() and GroovyShell.evaluate() methods, according to http://jira.codehaus.org/secu

20	0	src/test/groovy/ClosureInStaticMethodTest.groovy
4	0	src/test/groovy/script/HelloWorld2.groovy
0	10	src/test/groovy/script/TryStaticMethod.groovy
c333d4d994 Added AST level support for methods inside scripts. Just need the parser to be able to parse them
4	0	TODO.txt
4	0	src/main/org/codehaus/groovy/ast/MethodNode.java
21	1	src/main/org/codehaus/groovy/ast/ModuleNode.java
b8393d4b46 Added support for the running of Groovy scripts without a class required
5	0	TODO.txt
21	6	maven.xml
65	12	src/main/org/codehaus/groovy/ast/ModuleNode.java
4	0	src/main/org/codehaus/groovy/ast/stmt/BlockStatement.java
7	0	src/main/org/codehaus/groovy/syntax/lexer/StringCharStream.java
3	0	src/test/groovy/script/HelloWorld.groovy
3	0	src/test/groovy/script/PackageScript.groovy
10	0	src/test/groovy/script/TryStaticMethod.groovy
72	0	src/test/org/codehaus/groovy/ast/ModuleNodeTest.java
2	22	src/test/org/codehaus/groovy/syntax/parser/ASTBuilderTest.java
5	0	src/test/org/codehaus/groovy/syntax/parser/Parser

4df8b652aa fix possible deserialization exploit by overriding readObject
8	0	src/main/org/codehaus/groovy/runtime/MethodClosure.java

In [170]:
#f.	How many total lines of code (excluding comments and blank lines) 
#were deleted? Hint: git diff or git show
#Look at C

In [171]:
#g.	How many total lines of code (excluding comments and blank lines) 
#were added? Hint: git diff or git show
#Look at C

In [172]:
#h.	How many days were between the current fixing commit and the previous 
#commit of each affected file? Hint: git log
#AffectedFile
log_data = repo.git.log("--parents","--",affected_file).splitlines()
for line in log_data:
    print(line)

commit b25d0e55ddac82fa2127c69fb1ca10555975a2a5 91c04014fb6cc6a2e09f93d81fd0bd95ea84d7dc
Author: Cedric Champeau <cchampeau@apache.org>
Date:   Thu Dec 14 09:49:51 2017 +0100

    Move Java source set into `src/main/java`

commit 91c04014fb6cc6a2e09f93d81fd0bd95ea84d7dc 716d3e67e744c7edeed7cbc3f874090d39355764
Author: Daniel Sun <sunlan@apache.org>
Date:   Fri Nov 11 21:24:16 2016 +0800

    Implement the new parser

commit 716d3e67e744c7edeed7cbc3f874090d39355764 4c829ca397aae8f86c993c000d5ab8445c2a0189
Author: Jochen Theodorou <blackdrag@gmx.org>
Date:   Mon Oct 3 18:22:28 2016 +0200

    fix possible deserialization exploit by overriding readObject

commit 4c829ca397aae8f86c993c000d5ab8445c2a0189 09e9778e8a33052d8c27105aee5310649637233d
Author: oreissig <oreissig+github@gmail.com>
Date:   Sat Mar 12 22:28:30 2016 +0100

    make private fields final where possible (closes #292)

commit 09e9778e8a33052d8c27105aee5310649637233d 27234afb0182dfc43e6ebb68f8962680b526e1a8
Author: Cedric C

commit 716d3e67e744c7edeed7cbc3f874090d39355764 4c829ca397aae8f86c993c000d5ab8445c2a0189
Author: Jochen Theodorou <blackdrag@gmx.org>
Date:   Mon Oct 3 18:22:28 2016 +0200

    fix possible deserialization exploit by overriding readObject

commit 4c829ca397aae8f86c993c000d5ab8445c2a0189 09e9778e8a33052d8c27105aee5310649637233d
Author: oreissig <oreissig+github@gmail.com>
Date:   Sat Mar 12 22:28:30 2016 +0100

    make private fields final where possible (closes #292)
    
205 days

In [173]:
#i.	How many time has each affected file of the current fixing commit 
#been modified in the past since their creation (including rename of the file)? Hint: git log
#Affected File
log_data = repo.git.log("--oneline","--",affected_file).splitlines()
print("Number of commits in between:", len(log_data))
for line in log_data:
    print(line)


Number of commits in between: 26
b25d0e55dd Move Java source set into `src/main/java`
91c04014fb Implement the new parser
716d3e67e7 fix possible deserialization exploit by overriding readObject
4c829ca397 make private fields final where possible (closes #292)
09e9778e8a Fix potential problem when deserializing
27234afb01 remove "@version $Revision$" from javadoc
783f6d4080 license comment shouldn't be javadoc comment
a52d0d3c5d Update file headers for Apache
9992349e38 Optimize memory consuption by avoiding the creation of empty arrays
9ea9d18b6e GROOVY-4170 Fixed the issue related currying of private methods.
49c49791ae GROOVY-4104 MethodClosure didn't handle non-private, non-public methods of super classes correctly. Fixed it.
f2d80d59dd Support for optional arguments in closure meta methods and method-closure meta methods
f0ad471786 Fix for GROOVY-1993
6e20de5acb GROOVY-1710: Updated source file headers to ASL 2
469e823b00 modifications to ensure working method closure for static m

In [174]:
#j.	Which developers have modified each affected file since its creation? 
#Hint: git log
#AffectedFile
log_data = repo.git.log("--pretty=format:%an","--",affected_file).splitlines()
df = pd.DataFrame(log_data, columns=['Author'])
df['count'] = 1
answer = df.groupby(['Author']).count().reset_index(0, drop=False)
answer

Author  count
0       Alex Tkachman      2
1     Cedric Champeau      4
2          Daniel Sun      1
3   Guillaume Laforge      1
4      James Strachan      6
5    Jochen Theodorou      3
6         John Wilson      2
7           Paul King      1
8           Pilho Kim      2
9      Roshan Dawrani      2
10           oreissig      1
11   pascalschumacher      1

In [175]:
#k.	For each developer identified, how many commits have each of them 
#submitted? From your observation, are the involving developers
#experienced (with many commits) or new ones (with few commits) or both? 
#Hint: git log or git shortlog
commit_authors = repo.git.shortlog("-sne", "--all").splitlines()

author_commits = []

for commit_author in commit_authors:
    commit, author = commit_author.split("\t")
    author_commits.append([int(commit), author])
df = pd.DataFrame(author_commits, columns=['Commit', 'Author'])
df

Commit                                             Author
0      5858                     Paul King <paulk@asert.com.au>
1      2830               Jochen Theodorou <blackdrag@gmx.org>
2      2661        Cedric Champeau <cedric.champeau@gmail.com>
3      2135                James Strachan <jastrachan@mac.com>
4      1807                         sunlan <sunlan@apache.org>
5      1378             Guillaume Laforge <glaforge@gmail.com>
6      1306                         paulk <paulk@asert.com.au>
7       695           Roshan Dawrani <roshandawrani@gmail.com>
8       588                 Graeme Rocher <grocher@vmware.com>
9       491                     John Wilson <tug@wilson.co.uk>
10      466                     Daniel Sun <sunlan@apache.org>
11      443       John Wagenleitner <jwagenleitner@apache.org>
12      390        pascalschumacher <pascalschumacher@gmx.net>
13      373       Pascal Schumacher <pascalschumacher@gmx.net>
14      369            Danno Ferrin <danno.ferrin@shemnon.com>
15      355            Jeremy Rayner <jeremy.rayner@gmail.com>
16      340            Alex Tkachman <alex.tkachman@gmail.com>
17      332        PascalSchumacher <pascalschumacher@gmx.net>
18      306              Jason Dillon <jason.dillon@gmail.com>
19      276              Dierk Koenig <dierk.koenig@canoo.com>
20      275            danielsun1106 <realbluesun@hotmail.com>
21      236             Cedric Champeau <cchampeau@apache.org>
22      210                    Pilho Kim <phkim@cluecom.co.kr>
23      200                   Sam Pullara <sam@sampullara.com>
24      177            Franck Rasolo <franck.rasolo@gmail.com>
25      169               Russel Winder <russel@russel.org.uk>
26      161                   Boc McWhirter <bob@codehaus.org>
27      155                Hamlet D'Arcy <hamletdrc@gmail.com>
28      154                     Jim White <jim@pagesmiths.com>
29      147                  Thibault Kruse <kruset@in.tum.de>
..      ...                                                ...
370       1  Benjamin Porter <FreedomBen@users.noreply.gith...
371       1              Bruno Casali <brunoocasali@gmail.com>
372       1                     Christoph Frick <cf@ofnir.net>
373       1                         DJBen <lsh32768@gmail.com>
374       1        Dariusz Kobylarz <darek.kobylarz@gmail.com>
375       1                 David Avenante <altus34@gmail.com>
376       1               Gary Clayburg <gclaybur@comcast.net>
377       1                   Ingo Hoffmann <Ingo@Hoffmann.cx>
378       1                       Jakub Wilk <jwilk@jwilk.net>
379       1        James Williams <james.l.williams@gmail.com>
380       1         Jesper Steen Møller <jesper@selskabet.org>
381       1           Jess Sightler <jesse.sightler@gmail.com>
382       1                    Jesse Glick <jglick@apache.org>
383       1                Jochen Berger <fooberger@gmail.com>
384       1  Jochen Eddelbüttel <cornercouch@eddelbuettel.net>
385       1                         Joe <support@codehaus.org>
386       1                 Kohsuke Kawaguchi <kk@kohsuke.org>
387       1                        Larry J <larry@jalcyon.com>
388       1         Marcin Grzejszczak <marcin@grzejszczak.pl>
389       1  Marcin Zajaczkowski <szpak@users.sourceforge.net>
390       1                     Mark CC <support@codehaus.org>
391       1            Michael Frankfurter <micfra@micfra.net>
392       1                  Nathan Mische <nmische@gmail.com>
393       1                    Peter Swire <swirepe@gmail.com>
394       1                    Pierre Pinon <pierre@pinon1.fr>
395       1                  Roman Shaposhnik <rvs@apache.org>
396       1                  Wil Selwood <wselwood@ijento.com>
397       1                   Yasuharu NAKANO <ynak@jggug.org>
398       1                      aalmiray <aalmiray@yahoo.com>
399       1                          pavel <pavel@192.168.0.6>

[400 rows x 2 columns]

In [176]:
com = df[["Commit"]].copy()
df = pd.DataFrame(df.Author.str.split(' <',1).tolist(),columns = ['Author','email'])
df["Commit"] = com["Commit"]
names = answer["Author"].tolist()
for name in names:
    print(df[(df["Author"] == name)])

           Author                     email  Commit
16  Alex Tkachman  alex.tkachman@gmail.com>     340
             Author                       email  Commit
2   Cedric Champeau  cedric.champeau@gmail.com>    2661
21  Cedric Champeau       cchampeau@apache.org>     236
        Author                     email  Commit
10  Daniel Sun        sunlan@apache.org>     466
39  Daniel Sun  realbluesun@hotmail.com>      84
              Author                email  Commit
5  Guillaume Laforge  glaforge@gmail.com>    1378
           Author                email  Commit
3  James Strachan  jastrachan@mac.com>    2135
               Author                        email  Commit
1    Jochen Theodorou           blackdrag@gmx.org>    2830
296  Jochen Theodorou  jochen.theodorou@canoo.com>       2
        Author              email  Commit
9  John Wilson  tug@wilson.co.uk>     491
      Author                email  Commit
0  Paul King  paulk@asert.com.au>    5858
       Author                 email  Commi